File: Protocol/_theoreme_sim/src/routing/parentSelection.js
/* Purpose: Choose a strict higher-charge eligible parent using estimate-max and RSSI-like penalty. */

import { edgeKey } from "../core/graphUtils.js";

/**
 * @param {number} nodeId
 * @param {{
 *  nodes:Map<number,{id:number,qTotal:number,eligible:boolean,parent:number|null,isGateway:boolean}>,
 *  adjacency:Map<number,Set<number>>,
 *  estimates:Map<number,Map<number,number>>,
 *  edgeLookup:Map<string,{quality:number}>,
 *  config:{qForward:number,penaltyLambda:number,switchHysteresis:number}
 * }} state
 * @returns {number|null}
 */
export function chooseParent(nodeId, state) {
  const node = state.nodes.get(nodeId);
  if (node.isGateway || !node.eligible) {
    return null;
  }

  const neighbors = state.adjacency.get(nodeId) ?? new Set();
  const estimateMap = state.estimates.get(nodeId) ?? new Map();

  /** @type {Array<{id:number,penalty:number,score?:number,estimate:number}>} */
  const candidates = [];

  for (const neighborId of neighbors.values()) {
    const neighbor = state.nodes.get(neighborId);
    if (!neighbor || !neighbor.eligible) {
      continue;
    }

    // The theorem requires strict increase on real charges, not only estimates.
    if (neighbor.qTotal <= node.qTotal) {
      continue;
    }

    const estimate = estimateMap.get(neighborId) ?? 0;
    if (estimate <= node.qTotal || estimate < state.config.qForward) {
      continue;
    }

    const edge = state.edgeLookup.get(edgeKey(nodeId, neighborId));
    const quality = edge ? edge.quality : 0.1;
    const penalty = (1 - quality) * state.config.penaltyLambda;
    candidates.push({ id: neighborId, penalty, estimate });
  }

  if (candidates.length === 0) {
    return null;
  }

  const maxEstimate = Math.max(
    ...candidates.map((candidate) => candidate.estimate),
  );
  const tieWindow = Math.max(1, maxEstimate * 0.02);

  // Use penalty only among near-equal estimates as described in the theorem notes.
  const nearBest = candidates.filter(
    (candidate) => maxEstimate - candidate.estimate <= tieWindow,
  );

  for (const candidate of nearBest) {
    candidate.score = candidate.estimate - candidate.penalty;
  }

  nearBest.sort((a, b) => {
    if (b.score !== a.score) {
      return b.score - a.score;
    }
    if (b.estimate !== a.estimate) {
      return b.estimate - a.estimate;
    }
    return a.id - b.id;
  });

  const best = nearBest[0];
  const currentParent = node.parent;

  if (currentParent === null) {
    return best.id;
  }

  const current = nearBest.find((candidate) => candidate.id === currentParent);
  if (!current) {
    return best.id;
  }

  const keepCurrent =
    current.score + state.config.switchHysteresis >= best.score;
  return keepCurrent ? currentParent : best.id;
}

File: Protocol/_theoreme_sim_py/src/routing/parent_selection.py
"""Purpose: Select strict higher-charge eligible parent with hysteresis and link bonus."""

from __future__ import annotations

from src.propagation.link_usage_tracker import (
    compute_link_stability_bonus,
    get_effective_link_quality,
)


def choose_parent(node_id: int, state: dict) -> int | None:
    node = state["nodes"].get(node_id)
    if node is None or node.is_gateway or not node.eligible:
        return None

    neighbors = state["adjacency"].get(node_id, set())
    estimate_map = state["estimates"].get(node_id, {})
    candidates: list[dict] = []

    for neighbor_id in sorted(neighbors):
        neighbor = state["nodes"].get(neighbor_id)
        if neighbor is None or not neighbor.eligible:
            continue

        if float(neighbor.q_total) <= float(node.q_total):
            continue

        estimate = float(estimate_map.get(neighbor_id, 0.0))
        if estimate <= float(node.q_total) or estimate < float(state["config"]["qForward"]):
            continue

        quality = get_effective_link_quality(state, node_id, neighbor_id)
        penalty = (1 - quality) * float(state["config"]["penaltyLambda"])
        stability_bonus = compute_link_stability_bonus(state, neighbor_id, node_id)
        candidates.append(
            {
                "id": neighbor_id,
                "penalty": penalty,
                "estimate": estimate,
                "stabilityBonus": stability_bonus,
            }
        )

    if not candidates:
        return None

    max_estimate = max(candidate["estimate"] for candidate in candidates)
    tie_window = max(1.0, max_estimate * 0.02)

    near_best = [
        candidate
        for candidate in candidates
        if max_estimate - candidate["estimate"] <= tie_window
    ]

    for candidate in near_best:
        candidate["score"] = (
            candidate["estimate"] - candidate["penalty"] + candidate["stabilityBonus"]
        )

    near_best.sort(
        key=lambda item: (-item["score"], -item["estimate"], item["id"])
    )

    best = near_best[0]
    current_parent = node.parent

    if current_parent is None:
        return int(best["id"])

    current = next(
        (candidate for candidate in near_best if candidate["id"] == current_parent),
        None,
    )

    if current is None:
        current = next(
            (candidate for candidate in candidates if candidate["id"] == current_parent),
            None,
        )
        if current is not None:
            current["score"] = (
                current["estimate"]
                - current["penalty"]
                + current["stabilityBonus"]
            )

    if current is None:
        return int(best["id"])

    ratio = max(0.0, float(state["config"].get("switchHysteresisRatio", 0.0)))
    relative_margin = max(0.0, current["estimate"] * ratio)
    required_margin = max(float(state["config"]["switchHysteresis"]), relative_margin)

    keep_current = current["score"] + required_margin >= best["score"]
    return int(current_parent if keep_current else best["id"])
